### ISSUE TITLE
Stock Capacity and Storage

### ISSUE THREAD

**Issue by mokun** [mars-sim/mars-sim#496]
**Is your feature request related to a problem? Please describe.**
- Stock capacity is akin to the open storage space. One can leave whatever on the floor and without container.

**Describe the solution you'd like**
- The gases in a settlement should be a part of the stock storage because it doesn't need any container to hold these gases. On the other hand, one can have tanks of oxygen stored in a building in a settlement 
- Each building has predefined storage for certain amount resources but it also has a stock capacity 

**Additional context**
- MicroInventory has a map called storageMap. Inside that, it has specific `ResourceStored` object for each resource. 
- Currently, the idea of this general/stock storage space is not fully put into use. 

Q: how would we distinguish between storing a resource using stock storage space versus using a specific `ResourceStored`object ?

**Comment by mokun** [comment]
@bevans2000 

Before we proceed to working on the stock storage, I suppose it's a better to have two types of storage maps in MicroInventory, one for amount resources and the other one for item resources. 

 Do you see any issues with that in having separate maps ?

**Comment by mokun** [comment]
> ItemHolder - these can be just "placed" anywhere, this leads into your Storage issue here I think.

Yes. Let's create ItemHolder.

Can we make sure item resources (aka parts) make use of general storage when calculating capacity ?

> Settlements & Vehicles can do all 3. Person can only do Equipment & Items and lastly Containers just AmountResources.

In case of storing amount resources, there are 3 different approaches: 
- general/stock storage
- container storage
- specialized storage

Settlement & Vehicle use all 3 approaches.

Person and Robot use only the container storage approach.

EVA suit uses the specialized storage.

> GenericContainer has empty methods for holding ItemResources which always feels wrong.

Ok. we'll separate ItemHolder from ResourceHolder and GenericContainer won't need ItemHolder.

**Comment by mokun** [comment]
As you have suggested, it makes more sense to, say, store 100 kg of sesame oil into a barrel container (which in terms is stored in a settlement), instead of storing it directly in the shared capacity storage space of a settlement.

So, we need to apply the 2nd approach of using containers for storage in Settlement more often.

In the past, the `storeAmountResource()` method in Inventory will have a param asking if a container is needed.

Should we revive this approach ?

**Comment by bevans2000** [comment]
I'm not sure about that. For example, what about storing gas in a Vehicle or Settlement? Both of these would hold gases with containers because they have the capacity. But solids can be held directly without containers. So I dont think we need to add an extra parameters.

**Comment by mokun** [comment]
> For example, what about storing gas in a Vehicle or Settlement?

Both Settlement and Vehicle can use all 3 approaches for storing amount resources.

**Comment by mokun** [comment]
e.g. Inside each Rover, there are specialized storage for a number of resources.  But there is also a cargo storage for storing anything. There's also space for holding containers. 

Therefore, all 3 storage approaches can be applied to a Rover.

**Comment by mokun** [comment]
With regard to the storing amount resources in a rover, we have to decide what happens if, say, I want to store 1000 kg food in a vehicle, when the rover only have 128 kg for specialized food storage. 

Can the surplus of 872kg food be stored in the general/cargo/shared storage space of this rover ?

If it's true, then the `LoadingController` and some of those loading tests will have to change to accommodate this possibility, right ?

Realistically speaking, in mars-sim, storing an undetermined amount rock samples in the cargo storage is very legit.

**Comment by mokun** [comment]
Can you take a look at why this test in `LoadVehicleTest` won't pass ?
```java
	/*
	 * Test method for 'org.mars_sim.msp.simulation.person.ai.task.LoadVehicle.LoadingPhase(double)'
	 */
	public void testLoadRequiredAmountResources() throws Exception {
		Map<Integer, Number> requiredResourcesMap = new HashMap<>();
		requiredResourcesMap.put(ResourceUtil.foodID, 30D);
		requiredResourcesMap.put(ResourceUtil.waterID, 10D);

		// Load the manifest
		testLoading(200, requiredResourcesMap, Collections.emptyMap(),
					   Collections.emptyMap(), Collections.emptyMap());
	}
```

**Comment by bevans2000** [comment]
> Can you take a look at why this test in `LoadVehicleTest` won't pass ?
> 
> ```java
> 	/*
> 	 * Test method for 'org.mars_sim.msp.simulation.person.ai.task.LoadVehicle.LoadingPhase(double)'
> 	 */
> 	public void testLoadRequiredAmountResources() throws Exception {
> 		Map<Integer, Number> requiredResourcesMap = new HashMap<>();
> 		requiredResourcesMap.put(ResourceUtil.foodID, 30D);
> 		requiredResourcesMap.put(ResourceUtil.waterID, 10D);
> 
> 		// Load the manifest
> 		testLoading(200, requiredResourcesMap, Collections.emptyMap(),
> 					   Collections.emptyMap(), Collections.emptyMap());
> 	}
> ```

Sure. I'll have a look today

**Comment by bevans2000** [comment]
> With regard to the storing amount resources in a rover, we have to decide what happens if, say, I want to store 1000 kg food in a vehicle, when the rover only have 128 kg for specialized food storage.
> 
> Can the surplus of 872kg food be stored in the general/cargo/shared storage space of this rover ?
> 
> If it's true, then the `LoadingController` and some of those loading tests will have to change to accommodate this possibility, right ?
> 
> Realistically speaking, in mars-sim, storing an undetermined amount rock samples in the cargo storage is very legit.

I would say storing it in the storage area is fine. Where it breaks down is for Liquid and Gas, these would need containers if the volumes are greater than the specialised storage. I will have a look at this and the Loading  controller.

**Comment by bevans2000** [comment]
> > Can you take a look at why this test in `LoadVehicleTest` won't pass ?
> > ```java
> > 	/*
> > 	 * Test method for 'org.mars_sim.msp.simulation.person.ai.task.LoadVehicle.LoadingPhase(double)'
> > 	 */
> > 	public void testLoadRequiredAmountResources() throws Exception {
> > 		Map<Integer, Number> requiredResourcesMap = new HashMap<>();
> > 		requiredResourcesMap.put(ResourceUtil.foodID, 30D);
> > 		requiredResourcesMap.put(ResourceUtil.waterID, 10D);
> > 
> > 		// Load the manifest
> > 		testLoading(200, requiredResourcesMap, Collections.emptyMap(),
> > 					   Collections.emptyMap(), Collections.emptyMap());
> > 	}
> > ```
> 
> Sure. I'll have a look today

Problem here is due to the precision of doubles. In this test the Vehicle runs out of capacity before the loading completes even though the Vehicle capacity is set to the correct value at the start of the test. The root cause is the the double arithmetic starts to hit precision problems because the amount we can load each interaction is not an finite integer but is a double. 
The only way (& correct way) to solve this is to convert the Storage unit to be integers (grammes) instead of double (kilogrammes). Coding wise, this is an easy conversion BUT there are some calculations that will need adjusting, e.g. how Rock can be collected per second. Converting to grammes (& ints) would actually be better because integer arithmetic is faster.  I will create a dedicated Issue for this.
The short term fix I have done is to add 1% extra capacity to the Vehicle as part of the test so there is a little headroom.

**Comment by bevans2000** [comment]
> > With regard to the storing amount resources in a rover, we have to decide what happens if, say, I want to store 1000 kg food in a vehicle, when the rover only have 128 kg for specialized food storage.
> > Can the surplus of 872kg food be stored in the general/cargo/shared storage space of this rover ?
> > If it's true, then the `LoadingController` and some of those loading tests will have to change to accommodate this possibility, right ?
> > Realistically speaking, in mars-sim, storing an undetermined amount rock samples in the cargo storage is very legit.
> 
> I would say storing it in the storage area is fine. Where it breaks down is for Liquid and Gas, these would need containers if the volumes are greater than the specialised storage. I will have a look at this and the Loading controller.

Yes, I'll sort this out. First step is to add the concept of ```PhaseType``` to the container.

**Comment by mokun** [comment]
> I'm not sure about that. For example, what about storing gas in a Vehicle or Settlement?

I found out even though the gases in the atmosphere of Settlement are being tracked, they are not being stored in the stock capacity of the settlement yet.

But we could change that and have it updated per millisol.

Do you want these gases to be a part of the stock storage ?

**Comment by mokun** [comment]
Btw, we don't model the atmosphere in Vehicle yet but we could.

Also, talking about specialized storage, 

In a rover, there are cargo capacities as you know and they kind of act like specialized storage canisters. 

Of course, we do have specialized storage in some of the buildings in Settlement via the FunctionType.STORAGE 

But those specialized storage are NOT the same as the stock capacity. 

They are two different animals as intended.

**Comment by bevans2000** [comment]
So you are saying we don't model the consumption of oxygen for example in a settlement? Well that us not a function of the Settlement, it should be the responsibility of the LifeSupport function and its time pulse.

**Comment by mokun** [comment]
> So you are saying we don't model the consumption of oxygen for example in a settlement?

We do all the computation of gases in AirComposition class. But currently, I don't think we actually use Settlement's stock storage to store these gases. I'll need to take a look to be certain if we are using the **Stock Storage** in the settlement or if we are using **Specialized Storage** by `FunctionType.STORAGE`

What I'm saying is that **Stock Storage** is not the same as **Specialized Storage** (aka `FunctionType.STORAGE`)

**Comment by mokun** [comment]
@bevans2000 

We'll move this issue to 3.5.0 to fix.

**Comment by mokun** [comment]
@bevans2000 

When we start accounting for the payload mass of a rocket for returning home or going to the Moon, we can look into how to streamline the storage capacity of a settlement.
